addons/portal/controllers/mail.py

```diff
@@ -193,9 +214,16 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
             if not request.env['res.users'].has_group('base.group_user'):
                 domain = expression.AND([Message._get_search_domain_share(), domain])
             Message = request.env['mail.message'].sudo()
+
+        message_ids = Message.search(domain, limit=limit, offset=offset, order='date desc')
+        messages = message_ids.portal_message_format()
+        domain_next_attachment_ids = ['&', ('attachment_ids', '!=', False), ('id', 'not in', message_ids.ids)]
+        message_next_attachments_ids = Message.search(expression.AND([domain_next_attachment_ids, domain]))
```

Comment:
Mmmh ... strange ... you should cleanup that computation as it is a bit coding by ping-pong. You shoudl probably
1/ have IDs of all messages linked to the current record
2/ fetch a subset of those messages to return to the frontend (pagination)
3/ fetch all attachments linked to all messages

Here you fetch a window of messages, then other messages, then adding a method doing some attachments computation based on those two record sets, but a bit different due to access tokens. You should have a computation for a given model done at a single location ... For example your code make token generation for attachments harder to follow as it is now done in an additional place.

Reply:
I did this because token generation for attachments is (in the existing code) related to message formatting, and not attachment formatting. I could move this token generation as an optional part of the attachment formatting instead of the message formatting.

Also, with your method, we still have to avoid attachments linked to messages displayed in the current pagination, or else we will apply the attachment_format function twice.